openvpn: update to 2.7, support the new ovpn data channel offload module

LGA1150 opened the pull request:
## 📦 Package Details

**Maintainer:** @zhaojh329 @AuthorReflex @sch-m 

**Description:**
Update OpenVPN to 2.7, and use the new DCO module source.

---

## 🧪 Run Testing Details

- **OpenWrt Version:** r31013-19e9772935
- **OpenWrt Target/Subtarget:** mediatek/mt7622
- **OpenWrt Device:** Xiaomi Redmi Router AX6S

---

## ✅ Formalities

- [x] I have reviewed the [CONTRIBUTING.md](https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md) file for detailed contributing guidelines.

### If your PR contains a patch:

- [x] It can be applied using `git am`
- [x] It has been refreshed to avoid offsets, fuzzes, etc., using
  ```bash
  make package/<your-package>/refresh V=s
  ```
- [x] It is structured in a way that it is potentially upstreamable
<sub>(e.g., subject line, commit description, etc.)</sub>
<sub>We must try to upstream patches to reduce maintenance burden.</sub>

LGA1150 commented:
CI always fails at `(3/6) Installing liblz4-1` on mips_24kc. No idea what may cause it.

systemcrash commented:
Just curious: can those who use openvpn tell me why when WireGuard exists (clients for every platform, formally verified crypto)? Momentum? Any reason not to upgrade openvpn to v3? v3 seems a bit more complex and cumbersome.

sanderpqr commented:
I have noticed that with OpenVPN 2.6 and `kmod-ovpn-dco-v2`, some targets need their hardware crypto accelerators disabled (e.g., unload/remove `kmod-crypto-hw-safexcel` on `filogic`) for DCO to work. I haven't tested yet with this PR, but is this something others recognize either with or without this PR?

LGA1150 commented:
@sanderpqr Related issue: https://github.com/OpenVPN/ovpn-net-next/issues/14

sanderpqr commented:
>when using DCO, the crypto operations are handled by the kernel, which safexcel interacts with

Ah yes, this makes sense.

>Is safexcel used by any other application? It may well be that the module is simply broken on that platform.

I'm not aware of how to determine or test this. All I can say is that `kmod-crypto-hw-safexcel` is included for all `mt7623`, `filogic` and some `mvebu` targets. The `filogic` target is very popular, but the only issues I can find regarding the accelerator on GitHub and the forum relate to DCO...

On `mt7621` and `airoha` there is the related `EIP93` accelerator, I will do some tests this weekend with DCO on an mt7621-device to determine if similar issues exist with that combination.

LGA1150 commented:
> I'm not aware of how to determine or test this.

@sanderpqr You can test IPSec using the AES-GCM cipher.

LGA1150 commented:
@its0ka Thanks for your tests. Also as I stated in the [related issue](https://github.com/OpenVPN/ovpn-net-next/issues/14#issuecomment-2911795326), the driver passed all self-tests, so it's unlikely that the driver is broken.

its0ka commented:
@sanderpqr
I already had it removed because of issues on 24.10, installing it helped with reboots, but performance is worse now

ordex commented:
@LGA1150 @sanderpqr as for other kernel modules (i.e. batman-adv) it is always more convenient to use the out-of-tree code (like `ovpn-backports`), as it comes with the latest features.

Sticking to a specific kernel means never getting any new features, unless you do a full kernel upgrade.

LGA1150 committed:
ovpn-dco: switch to the new source

The ovpn kernel module is is part of the Linux kernel starting from
version 6.16.
Switch to the new officially maintained backports source:
https://github.com/OpenVPN/ovpn-backports

Signed-off-by: Qingfang Deng <dqfext@gmail.com>

LGA1150 committed:
openvpn: update to 2.7.1

The new DCO module depends on OpenVPN 2.7.1.
For details refer to https://github.com/OpenVPN/openvpn/blob/v2.7.1/Changes.rst

Removed upstreamed wolfSSL patches:
- 101-Fix-EVP_PKEY_CTX_-compilation-with-wolfSSL.patch
- 102-Disable-external-ec-key-support-when-building-with-wolfSSL.patch

Reworked 100-mbedtls-disable-runtime-version-check.patch to use
MBEDTLS_VERSION_STRING instead of a mutable buffer.

Signed-off-by: Qingfang Deng <dqfext@gmail.com>

ptpt52 reviewed (approved):
LGTM

hnyman commented:
@systemcrash wrote:
> I recommend waiting until the proto handler migration stabilises some more - 2.7.x removed a number of deprecated options and likely requires changes there to adapt. Also, the first release usually contains a few bugs, so perhaps .1 would be a better

Stabilised enough?

systemcrash commented:
Yeah, LGTM @hnyman

hnyman commented:
Let's then merge this, so that the other PRs for openvpn can be rebased on top.

ordex commented:
Congrats everybody for getting this in!
I'm so excited to see OpenVPN 2.7 and the new oven kernel module on OpenWrt!

ordex commented:
@LGA1150 may it make sense to force safexcel off, when kmod-ovpn is selected?

sanderpqr commented:
I think we first need some more troubleshooting to have an idea of the cause. For now, only DCO seems to have issues with the SafeXcel accelerator. That said, the accelerator-code isn't optimal, see the efforts to get the related `EIP93` working properly for example: https://github.com/openwrt/openwrt/pull/20130

abajk commented:
> I have noticed that with OpenVPN 2.6 and `kmod-ovpn-dco-v2`, some targets need their hardware crypto accelerators disabled (e.g., unload/remove `kmod-crypto-hw-safexcel` on `filogic`) for DCO to work. I haven't tested yet with this PR, but is this something others recognize either with or without this PR?

@sanderpqr Safexcel is broken on MT7988. What device has this issue?

sanderpqr commented:
> Safexcel is broken on MT7988. What device has this issue?

@abajk I've tested this on Cudy TR3000 (MT7981), ZyXEL T-56 and GL-MT6000 (both MT7986). The issue persists with kernel 6.18 + OpenVPN 2.7.1 with `kmod-ovpn-backports`

ralflici commented:
I also tested ovpn on OpenWrt 6.12 using GL.iNet GL‑MT6000 (MT7986) with AES‑GCM and compared `crypto_safexcel` against software/SIMD providers. In this setup, safexcel was consistently slower and significantly less stable than software/SIMD, with the largest issues appearing on TCP transport. Safexcel appears to rely heavily on SWIOTLB bounce buffering under ovpn load, which adds copy/allocation overhead and likely contributes materially to the throughput gap versus software/SIMD.

On UDP, dmesg frequently showed:
```text
crypto-safexcel 10320000.crypto: swiotlb buffer is full (sz: 8 bytes), total 512 (slots), used 512 (slots)
crypto-safexcel 10320000.crypto: swiotlb buffer is full (sz: 16 bytes), total 512 (slots), used 512 (slots)
```
This indicates SWIOTLB pressure/exhaustion (512 slots observed on this platform/boot). Enabling `CONFIG_SWIOTLB_DYNAMIC` removed these warnings in my tests and improved throughput.

On TCP transport, the issue was worse as traffic would stall with repeated:
```text
tun0: unsupported protocol received from peer 0
```
This message is emitted after the decrypt path, so it means the resulting inner packet bytes are not valid L3 payload for ovpn.

I then tested an ovpn‑side workaround for the TCP + safexcel decrypt path (on top of `CONFIG_SWIOTLB_DYNAMIC`) by using private scratch memory (separate source/destination handling) and copying the plaintext back into the skb with `skb_store_bits` after completion. This removed the catastrophic RX behavior and made TCP runs usable. However, this is still only a mitigation, not a real fix: it does not address the underlying issue (which appears to be in safexcel from what I can see), and it adds extra per‑packet copy cost plus additional temporary memory pressure on top of safexcel’s own operations.

LGA1150 commented:
@ralflici I don't know why swiotlb is being used. The EIP-197 should be able to DMA all the 64-bit address space:

https://github.com/torvalds/linux/blob/v6.12/drivers/crypto/inside-secure/safexcel.c#L1782

Could you disable swiotlb by passing `swiotlb=noforce` to kernel cmdline and see what happens?

ralflici commented:
@LGA1150 yes, good suggestion. I just ran a quick iperf3 test with `swiotlb=noforce`, and both UDP and TCP now work without any warnings. Throughput is generally good, but there is noticeable per‑second variance and spikes of retransmissions compared to the non‑safexcel runs.

However `swiotlb=noforce` is documented as a debug mode so the question now is why exactly safexcel uses swiotlb in this setup.

ralflici commented:
I think the key point is that `dma_set_mask_and_coherent(..., DMA_BIT_MASK(64))` in safexcel only guarantees addressability, not that it would neven bounce. My understaing is that the decision to bounce is later in the DMA mapping path (`dma_map_sg` -> `dma_direct_map_phys`), and on arm64 we also have `CONFIG_DMA_BOUNCE_UNALIGNED_KMALLOC=y`. In the non-coherent case, small `FROM_DEVICE`/`BIDIRECTIONAL` mappings can still hit the bounce heuristics (`dma_kmalloc_needs_bounce` path), even with a 64-bit DMA mask.

For ovpn AEAD decrypt, this matters because we currently use in-place mapping (`src == dst`, so `DMA_BIDIRECTIONAL`) and the SG includes very small entries (AAD/tag). That matches the warnings I was seeing (`sz: 8` / `sz: 16`).

A possible mitigation on ovpn side (instead of globally disabling SWIOTLB) might be to keep normal 3-entry source SG (`AAD|payload|tag`) as input, use a single linear destination buffer for decrypt output (out-of-place) then copy plaintext payload back to skb. This avoids tiny `FROM_DEVICE/BIDIRECTIONAL` destination mappings while keeping protocol semantics. It’s still a workaround (extra copy), but it may reduce the pathological bounce behavior without requiring `swiotlb=noforce`.

ordex commented:
> It’s still a workaround (extra copy), but it may reduce the pathological bounce behavior without requiring `swiotlb=noforce`.

wouldn't this still produce lower performance compared to the software case (i.e. without safexcel)?

LGA1150 commented:
Is the EIP-197 or the ethernet controller on MT7986 DMA-coherent? The ethernet controller on MT7622 is, as indicated by the `dma-coherent` property in its device tree. But such property is missing in MT7986. You may try adding one to the EIP-197 node.

LGA1150 commented:
> use a single linear destination buffer for decrypt output (out-of-place) then copy plaintext payload back to skb.

When you do out-of-place you can just build a new skb on the linear buffer.

ralflici commented:
> wouldn't this still produce lower performance compared to the software case (i.e. without safexcel)?

Most likely.

> When you do out-of-place you can just build a new skb on the linear buffer.

Right, that would avoid the copy, but we would still need to allocate a new skb for the output, no?

> Is the EIP-197 or the ethernet controller on MT7986 DMA-coherent? The ethernet controller on MT7622 is, as indicated by the dma-coherent property in its device tree. But such property is missing in MT7986. You may try adding one to the EIP-197 node.

This is interesting, thanks. I'll see if it works.

abajk commented:
> Is the EIP-197

The MT7986 has EIP-97 built in. As far as I know, it's not coherent on this platform.

LGA1150 commented:
> Right, that would avoid the copy, but we would still need to allocate a new skb for the output, no?

True, but the SLUB allocator can reuse recently freed skbs from its per cpu cache, and you're freeing the old skb anyway.

ralflici commented:
> The MT7986 has EIP-97 built in. As far as I know, it's not coherent on this platform.

Indeed. I tried adding `dma-coherent` to the crypto node in the device tree and no tunnel traffic flowed at all.

> True, but the SLUB allocator can reuse recently freed skbs from its per cpu cache, and you're freeing the old skb anyway.

Sure, my concern was mostly that we should not make the normal software/SIMD path pay extra allocations or copies, since that path is already fast and does not show this failure mode. If this stays limited to safexcel, then that concern is much smaller.

I also tried the simpler out-of-place idea that decrypts safexcel traffic into a separate scratch destination and then copies the plaintext back with `skb_store_bits()` as a first test. With that change, the previous safexcel symptoms disappeared in my GL-MT6000 tests: no `swiotlb buffer is full` messages and no `unsupported protocol` errors. TCP also became usable instead of hitting the previous bad decrypt/corruption behavior. Performance is still generally below non-safexcel mode but it does seem to confirm that the large failure mode is related to the decrypt-side destination mapping/bounce behavior.
